### PR TITLE
[DPTP-1731] add the ability to create rbacs and wait until the image pull secret has been created

### PR DESCRIPTION
**Pull request by droslean**
When a service account is being created, the [DockercfgController](https://github.com/openshift/openshift-controller-manager/blob/master/pkg/serviceaccounts/controllers/create_dockercfg_secrets.go#L315) creates the pull secret and then update the service account. Therefore if the service account will be used before the creation of that secret, the running pod will pull images as anonymous instead. 

This PR fixes this race condition by creating a new function that will be responsible for the RBAC creation and waits until the dockercfg controller will create the secret and sync the service account.

Jobs with the related error: [ci-search - Container .* is not ready with reason ImagePullBackOff](https://search.ci.openshift.org/?search=Container+.*+is+not+ready+with+reason+ImagePullBackOff&maxAge=24h&context=0&type=build-log&name=.*&maxMatches=1&maxBytes=&groupBy=job)

/cc @openshift/openshift-team-developer-productivity-test-platform 

Signed-off-by: Nikolaos Moraitis <nmoraiti@redhat.com>

### CODE REVIEW

**Comment by alvaroaleman** [comment]
> This PR fixes this race condition by creating a new function that will be responsible for the RBAC creation and waits until the dockercfg controller will create the secret and sync the service accoun

Why not simply poll until the Secret is created? The approach taken here seems overly complicated.

**Comment by droslean** [comment]
> Why not simply poll until the Secret is created? The approach taken here seems overly complicated.

Well, we don't know the secret name to start with.  Also, this approach isn't complicated. We just wait until our service account has a value in its `imagePullSecrets`

File: pkg/util/rbacs.go [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+		return results.ForReason("creating_service_account").WithError(err).Errorf("could not create service account '%s'", sa.Name)
+	}
+	if err := client.Create(ctx, role); err != nil && !kerrors.IsAlreadyExists(err) {
+		return results.ForReason("creating_roles").WithError(err).Errorf("could not create role '%s'", role.Name)
+	}
+	if err := client.Create(ctx, roleBinding); err != nil && !kerrors.IsAlreadyExists(err) {
+		return results.ForReason("binding_roles").WithError(err).Errorf("could not create role binding '%s'", roleBinding.Name)
+	}
+
+	if err := retry.OnError(
```
Comment (alvaroaleman) [inline pkg/util/rbacs.go:29]: This is a lot simpler with `k8s.io/apimachinery/pkg/util/wait.Poll`

**Comment by alvaroaleman** [comment]
> Well, we don't know the secret name to start with. Also, this approach isn't complicated. We just wait until our service account has a value in its imagePullSecrets

Ah sorry, I was confused by the PRs title and didn't understand why we create RBAC now, but its just code that got moved around

File: pkg/steps/multi_stage.go [hunk 2]
```diff
@@ -240,18 +241,11 @@ func (s *multiStageTestStep) setupRBAC() error {
...
-	}
-	if err := s.client.Create(context.TODO(), sa); !check(err) {
-		return err
-	}
-	if err := s.client.Create(context.TODO(), role); !check(err) {
-		return err
-	}
-	if err := s.client.Create(context.TODO(), binding); !check(err) {
+
+	if err := util.CreateRBACs(context.TODO(), sa, role, binding, s.client, 1*time.Minute); err != nil {
```
Comment (alvaroaleman) [inline pkg/steps/multi_stage.go:245]: Why did you move that to a different package when its only used here? And its kind of confusing to have a `CreateRBAC` func that requires a ServiceAccount as argument. Also, doesn't the ServiceAccount poll code belong into `InitializeNamespace`?

File: pkg/util/rbacs.go (hunk 1 above)
Comment (droslean) [reply pkg/util/rbacs.go:29]: Thanks. Let me check that.

File: pkg/steps/multi_stage.go (hunk 2 above)
Comment (alvaroaleman) [reply pkg/steps/multi_stage.go:245]: Ah, its also used in CreateRelease. Still, I think the part that needs factoring out is the SA creation + Polling, combining it with the RBAC just makes it less re-usable and more confusing

Comment (droslean) [reply pkg/steps/multi_stage.go:245]: There are currently two places that we are creating rbacs. Multi-stage and create_release. I moved it to `util` package because it makes sense to have this kind of functionality in a neutral place. In `initializeNamespace` for some reason, we don't create any related rbacs.

Comment (droslean) [reply pkg/steps/multi_stage.go:245]: yea. Also, they are creating different serviceaccounts. The whole concept here needs some refactoring for sure.

File: pkg/util/rbacs.go [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+	}
+
+	if err := client.Create(ctx, role); err != nil && !kerrors.IsAlreadyExists(err) {
+		return results.ForReason("creating_roles").WithError(err).Errorf("could not create role '%s'", role.Name)
+	}
+	if err := client.Create(ctx, roleBinding); err != nil && !kerrors.IsAlreadyExists(err) {
+		return results.ForReason("binding_roles").WithError(err).Errorf("could not create role binding '%s'", roleBinding.Name)
+	}
+
+	if !skipPoll {
```
Comment (alvaroaleman) [inline pkg/util/rbacs.go:39]:
```suggestion
	if skipPoll {
	   return nil
	}
```

See https://blog.codinghorror.com/flattening-arrow-code/ for some samples on how to avoid this kind of arrow code

Comment (droslean) [reply pkg/util/rbacs.go]: addressed

**Review by stevekuznetsov — APPROVED** [review]
